### ISSUE TITLE
[RFE] Implement Remaining Terrains and Terrain Modifications

### ISSUE THREAD

**Issue by stonewall072** [MegaMek/megamek#6303]
### Prerequisites and Pre-Issue Checklist

- [X] I'm reporting the RFE to the correct repository:
- [MegaMek](https://github.com/MegaMek/megamek/issues)
- [MegaMekLab](https://github.com/MegaMek/megameklab/issues)
- [MekHQ](https://github.com/MegaMek/mekhq/issues)

- [X] Does this conform with our Content Policy linked above?
- [X] I've searched the GitHub tracker and haven't found a similar feature request listed

### RFE Details *

This is an improvement to an existing feature

### Brief Description *

The following terrains and terrain modifications still need implementation in MegaMek:

_TacOps AR:_

- Gravel Piles 
- Half Levels
- Bug Storms
- Water Flow
- Extreme Depths
- Hazardous Liquid Pools
- Rails
- Earthquake
- Meteor Showers
- Tainted and Toxic Atmospheres
- Additional building features such as doors and hexside walls, tents, fences, Castles Brian 

_BattleMech Manual:_

- Mud (Simplified) 

_Map Pack Alien Worlds:_

- Atmosphereic Domes (w/ Explosive Decompression)
- Ultra Sublevels
- Fungal Spores
- Crystal

_Battle of Tukayyid:_

- Improved Fieldworks
- Steep Cliffs (I *think* this is just Sheer Cliffs from TO:AR with a different name)
- Unstable 

_BattleTech Essentials:_

- Hardpacked Gravel

### MegaMekLab Suite Version *

0.50.02

### Final Checklist

- [X] I've checked to make sure that this RFE has not already been filed
- [X] I am requesting implementation of only one relevant set of features in this ticket for clarity and focus

**Comment by SimonLandmine** [comment]
The _Improved Fieldworks_ from _Battle of Tukayyid_ are a slight, weirdly directional, variation on our existing improved infantry positions, I believe, although it's been a while since I've looked at them.

**Comment by stonewall072** [comment]
> The _Improved Fieldworks_ from _Battle of Tukayyid_ are a slight, weirdly directional, variation on our existing improved infantry positions, I believe, although it's been a while since I've looked at them.

Yeah, it certainly looks that way, says it uses the same rules as fieldworks but only if attacks come through the marked hex side. 

Which also reminds me to add the actual weird pseudo-building "improved positions" to the list.

**Comment by HammerGS** [comment]
I can confirm Steep Cliffs (I think this is just Sheer Cliffs from TO:AR with a different name) is Sheer Cliffs with a tweak. We can use what we have.

**Comment by psikomonkie** [comment]
Hey Artists! I'd like to implement the following terrains first, if you can prioritize getting this art that'd be wonderful:

Highest Priority:
- [ ] Crystals
- [ ] Hazardous Liquid Pools

Other Priorities:
- [ ] Ultra Sublevels
- [ ] Hard Packed Gravel
- [ ] Unstable Terrain

Thanks,
Psi

**Comment by Morkath** [comment]
![ImprovedPosition](https://github.com/user-attachments/assets/f731c07f-84c8-4b04-a002-18226143d561)

Here is some art for an improved position.

**Comment by Morkath** [comment]
[Crystals.zip](https://github.com/user-attachments/files/18304859/Crystals.zip)
![crystal_diamond](https://github.com/user-attachments/assets/1828de1d-9d31-4e2d-a9f4-955d1c2cfa9b)
Crystal terrain, @psikomonkie I recommend using the diamond (pictured) as the default as I think it turned out the best.

**Comment by Morkath** [comment]
[Crystals.zip](https://github.com/user-attachments/files/18305035/Crystals.zip)
![crystal_diamond](https://github.com/user-attachments/assets/dbe5dca0-b21f-4370-a951-c2bd6d24dac4)
![crystal_amber](https://github.com/user-attachments/assets/76256321-65b1-4a0d-94d9-7d73952c7541)
![crystal_amethyst](https://github.com/user-attachments/assets/22fb5ba1-63e8-4b44-9405-350b73625d6b)
![crystal_emerald](https://github.com/user-attachments/assets/5f7fab65-fb3f-453d-9287-e1bb061ae4d1)
![crystal_ruby](https://github.com/user-attachments/assets/4fe04281-a662-4669-9c63-af90917aff01)
![crystal_sapphire](https://github.com/user-attachments/assets/63a6ec03-dbf8-4e51-92a3-4bd431120b40)

Crystal terrain.  @psikomonkie the diamond, ruby, and emerald turned out the best.  I would recommend using the diamond as the default, followed by ruby/emerald, then the others.

**Comment by stonewall072** [comment]
[enhanced_gravel_road.zip](https://github.com/user-attachments/files/18305385/enhanced_gravel_road.zip)

I went through and made some additional adjustments to the gravel road images to improve their look a bit.

![gravel_road00](https://github.com/user-attachments/assets/3b49ae5b-cbd1-4158-877a-046cecb9315d)

**Comment by Morkath** [comment]
[Toxic_Water.zip](https://github.com/user-attachments/files/18312657/Toxic_Water.zip)
![yellow_toxic_water_blobs](https://github.com/user-attachments/assets/634aa99e-4cb4-4e19-9791-8fd386d81ffc)
![green_toxic_water_blobs](https://github.com/user-attachments/assets/65494f5a-697a-4e6b-ba9b-18c813235b17)
![green1_toxic_vats](https://github.com/user-attachments/assets/e46ed0af-c976-4539-9c05-878d5fa22301)
![green2_toxic_vats](https://github.com/user-attachments/assets/d8616296-d49d-4d9f-a759-ee88b3601a13)
![green3_toxic_vats](https://github.com/user-attachments/assets/a9c28187-2f56-46e8-a9e2-9e125622ef45)

Some art for hazardous liquid pools.

**Comment by Morkath** [comment]
[toxic_water2.zip](https://github.com/user-attachments/files/18312688/toxic_water2.zip)
![green1_toxic_water](https://github.com/user-attachments/assets/457aaf8a-2bd1-4278-8432-ade57602d49e)
![green2_toxic_water](https://github.com/user-attachments/assets/da691077-babb-4e8f-b8d8-1881a3c018e3)
Oh and some full hex versions if you want them.

**Comment by Morkath** [comment]
@stonewall072 @psikomonkie I don't think we need specific art for unstable, looking at the circle of equals map it should just be a modifier that can be applied on any rubble, and puts "Unstable" text on the hex like CGL did on CoE.

**Comment by stonewall072** [comment]
Hardpacked Gravel

![hardpackedgravel](https://github.com/user-attachments/assets/428ec251-ade8-4d9f-85e5-10c37ca7018d)

**Comment by stonewall072** [comment]
![gravelpile5](https://github.com/user-attachments/assets/8d7edc7b-42e5-4673-bb65-abc7e35fbfcb)

Gravel Pile. Future enhancement if and when the terrain modification is implemented would be utilizing the pattern with something similar to water hexes to create more "flowing" edges to the piles.

**Comment by Morkath** [comment]
![green2_toxic_water_blobs](https://github.com/user-attachments/assets/0421c759-5bbf-4294-9af6-93d65690fb3b)
![purple_toxic_water_blobs](https://github.com/user-attachments/assets/e692bbf9-8558-4558-b3c2-65fd3ab7f9aa)
![purple2_toxic_water_blobs](https://github.com/user-attachments/assets/5285657a-de00-4875-9b3d-566cd4270956)
[toxic_water3.zip](https://github.com/user-attachments/files/18324444/toxic_water3.zip)
![yellow2_toxic_water_blobs](https://github.com/user-attachments/assets/4e330775-fc9a-4695-aeab-ed7053a91e12)

A few more variations for hazardous pools.

**Comment by Morkath** [comment]
![bug_swarm](https://github.com/user-attachments/assets/efcb78ee-97cb-4621-9902-20114746c5a6)
bug swarm sprite

**Comment by Morkath** [comment]
![ultra_sublevel](https://github.com/user-attachments/assets/65ca806f-85bb-4fe3-b550-1dcd0c7bb540)
ultra sublevel, this should probably be replaced by a better artist, but if you don't get anything else it should work for now

**Comment by Morkath** [comment]
[improved_fieldworks.zip](https://github.com/user-attachments/files/18353076/improved_fieldworks.zip)
![improved_fieldworks1](https://github.com/user-attachments/assets/09c0ac37-59a4-4924-ab6e-119866c35b14)
![improved_fieldworks2](https://github.com/user-attachments/assets/85ae7a59-b8eb-4c81-a2a3-7e74e4908c7a)
![improved_fieldworks3](https://github.com/user-attachments/assets/6f3b82b9-447e-4b08-a21e-818b071915dd)
![improved_fieldworks4](https://github.com/user-attachments/assets/d6ca4611-2343-4a65-9bea-d3dae8342251)
![improved_fieldworks5](https://github.com/user-attachments/assets/bf1bd428-3a63-466a-87e2-ffeb7f0a3838)
![improved_fieldworks6](https://github.com/user-attachments/assets/726b1e6a-1751-4680-8d4d-752afbef3ed5)

Improved fieldworks

**Comment by Morkath** [comment]
![hot_mud](https://github.com/user-attachments/assets/28cb3c54-8fa2-4f57-8320-6a9e28f20c6a)
Hot mud

**Comment by Morkath** [comment]
[meteor_swarm.zip](https://github.com/user-attachments/files/18353433/meteor_swarm.zip)
![meteor_swarm_single1](https://github.com/user-attachments/assets/ea9b45c6-dd39-4274-b78b-19c57332c175)
![meteor_swarm_single2](https://github.com/user-attachments/assets/b4cb47a2-340c-43b1-ac6e-3656841efaf5)
![meteor_swarm_single3](https://github.com/user-attachments/assets/a33351b8-aec2-4cf9-8924-b2c08bf6f02d)
![meteor_swarm_single4](https://github.com/user-attachments/assets/f90e7447-c47f-4a1c-a9fe-cab8ed97be78)

Meteor Swarm impact indicators

**Comment by Morkath** [comment]
![blue_tent](https://github.com/user-attachments/assets/380e9f4a-3806-44c0-a7dc-e0b0e75c0935)
![blue_tents_multiple](https://github.com/user-attachments/assets/de556946-be14-4704-824c-06ae435996fd)
![green_tent](https://github.com/user-attachments/assets/7d982fe0-d671-4136-9140-9dbe4bbba792)
![green_tent_multiple](https://github.com/user-attachments/assets/3cf63a20-4017-404c-8039-b0fd718b03b2)
![red_tent](https://github.com/user-attachments/assets/02087ac6-fba2-437b-82be-31fae86a7013)
![red_tents_multiple](https://github.com/user-attachments/assets/41da5db2-c99b-4239-b091-5da6732e82c9)
[tents.zip](https://github.com/user-attachments/files/18355818/tents.zip)
Some tents.

**Comment by Morkath** [comment]
![green_tent_multiple](https://github.com/user-attachments/assets/2f275c89-f3a1-4006-8d12-abb612c2eadb)
![red_tent2](https://github.com/user-attachments/assets/f1dc98b5-c45c-4170-a0e7-fc54a711b8cb)
[tents2.zip](https://github.com/user-attachments/files/18356051/tents2.zip)
![blue_tent2](https://github.com/user-attachments/assets/84481f9b-6bdc-4dab-91b9-fad2afdc3065)
Realized people might want big versions of the tents without the fluff additions for other environments.

**Comment by Morkath** [comment]
![blue_tent2](https://github.com/user-attachments/assets/5638bd39-b6f3-4519-ab80-3168a6f094f1)
![green_tent2](https://github.com/user-attachments/assets/0aa4a5bd-a478-41fe-8918-6c93fd04b52b)
![red_tent2](https://github.com/user-attachments/assets/84347610-c68c-4dad-a265-28783f4774d1)
[tents2.zip](https://github.com/user-attachments/files/18356054/tents2.zip)
Realized people might want versions of the big tents without the fluff additions, for other environments.

**Comment by Morkath** [comment]
[fence.zip](https://github.com/user-attachments/files/18356337/fence.zip)
![fence1](https://github.com/user-attachments/assets/2348227c-8dbd-487b-876a-71c003bcd12c)
![fence2](https://github.com/user-attachments/assets/e0c6fb6f-79ff-49e2-9c2b-79bf4ee4cdbc)
![fence3](https://github.com/user-attachments/assets/730957f0-edbc-47c6-b404-de9fa3ef4138)
![fence4](https://github.com/user-attachments/assets/bea68105-e60a-44df-a175-ba8f849bd593)
![fence5](https://github.com/user-attachments/assets/f25a1811-4f73-406a-9e56-93af95a3a0e2)
![fence6](https://github.com/user-attachments/assets/9c1d6bad-bb7f-41e8-a15b-35987b9214e6)
fence

**Comment by psikomonkie** [comment]
Hazardous Liquid Pools do not match the saxarba's tileset's water, I think this would be best if we had a new set of "Structured_Water" for Hazardous Liquid that's a new color from the other saxarba waters. To match the saxarba water it'll need images for 5 depths and the 64 different "exit" patterns.

Thanks,
Psi

**Comment by stonewall072** [comment]
> Hazardous Liquid Pools do not match the saxarba's tileset's water, I think this would be best if we had a new set of "Structured_Water" for Hazardous Liquid that's a new color from the other saxarba waters. To match the saxarba water it'll need images for 5 depths and the 64 different "exit" patterns.
> 
> Thanks, Psi

Probably be easiest to just color shift the existing set. To that end, could we use the Mars theme water as a stop gap? It'd be sufficiently different enough *on most maps*.

**Comment by psikomonkie** [comment]
> > Hazardous Liquid Pools do not match the saxarba's tileset's water, I think this would be best if we had a new set of "Structured_Water" for Hazardous Liquid that's a new color from the other saxarba waters. To match the saxarba water it'll need images for 5 depths and the 64 different "exit" patterns.
> > Thanks, Psi
> 
> Probably be easiest to just color shift the existing set. To that end, could we use the Mars theme water as a stop gap? It'd be sufficiently different enough _on most maps_.

So that was my original plan - but the other themes don't look to have the 64 different exit patterns - the Mars, Volcano, Lunar, etc. only have the different depths and don't support the fancy sides.

**Comment by stonewall072** [comment]
@psikomonkie 

[StructuredLiquids.zip](https://github.com/user-attachments/files/18631183/StructuredLiquids.zip)

Attached is a zip with the structured water hexes recolored for use with hazardous liquids and themes. Filenames got mangled in the process, but order should be the same as what's in the structured water folder. Nothing someone with some skill can't handle programmatically ;).

**Comment by psikomonkie** [comment]
> [@psikomonkie](https://github.com/psikomonkie)
> 
> [StructuredLiquids.zip](https://github.com/user-attachments/files/18631183/StructuredLiquids.zip)
> 
> Attached is a zip with the structured water hexes recolored for use with hazardous liquids and themes. Filenames got mangled in the process, but order should be the same as what's in the structured water folder. Nothing someone with some skill can't handle programmatically ;).

The order is not the same as the structured water folder. Image  "1" corresponds to "1", which is right, but then image "02" corresponds with "10" and "03" corresponds to "11". That pattern holds true for the first two or three groups of 64, but then the pattern changes again - 194, the first of the last group of 64, corresponds to "64", its 2nd corresponds to "55", and the third corresponds to "54" - which torpedoed my idea of remapping it with powershell.

**Comment by HammerGS** [comment]
See here for  Crystalline Canyon - https://bg.battletech.com/forums/index.php?topic=87209.msg2057412#msg2057412

Earthquake on 10+

**Comment by Morkath** [comment]
[heavy_wall.zip](https://github.com/user-attachments/files/19221505/heavy_wall.zip)
![Image](https://github.com/user-attachments/assets/1ffb0ae3-1d64-4495-9ff8-194142cc6a65)

Heavy version of the hexside walls.

Still needs light, medium, and reinforced.

**Comment by Morkath** [comment]
[light_wall.zip](https://github.com/user-attachments/files/19269705/light_wall.zip)
[medium_wall.zip](https://github.com/user-attachments/files/19269707/medium_wall.zip)
[hardened_wall.zip](https://github.com/user-attachments/files/19269708/hardened_wall.zip)

[armored_light_wall.zip](https://github.com/user-attachments/files/19269709/armored_light_wall.zip)
[armored_medium_wall.zip](https://github.com/user-attachments/files/19269710/armored_medium_wall.zip)
[armored_heavy_wall.zip](https://github.com/user-attachments/files/19269711/armored_heavy_wall.zip)
[armored_hardened_wall.zip](https://github.com/user-attachments/files/19269714/armored_hardened_wall.zip)

![Image](https://github.com/user-attachments/assets/7b7dd98d-2219-41cf-bdf7-c442c83c4294)
![Image](https://github.com/user-attachments/assets/e61e1f1e-0248-49a6-a534-cc1d574edd43)
![Image](https://github.com/user-attachments/assets/c51f8a4e-8868-4867-bddd-c714b227c6c7)

![Image](https://github.com/user-attachments/assets/8060fd1a-a03f-405c-9887-8428d74e7a01)
![Image](https://github.com/user-attachments/assets/1c7012ca-3db9-418a-bdae-8c07cf4d9213)
![Image](https://github.com/user-attachments/assets/09e3bbfa-8cb1-4946-8b9c-f0794625fd9e)
![Image](https://github.com/user-attachments/assets/13c1d5ff-9164-4183-a87f-962618a8daf7)

Light, Medium, and Hardened Walls.
Light, Medium, Heavy, and Hardened Armored Walls.

**Comment by Morkath** [comment]
And that should be the last of what I can contribute, art still needed is beyond my skill and by my reckoning is;

- Atmosphereic Domes (w/ Explosive Decompression)
- Castles Brian

**Comment by SimonLandmine** [comment]
Quick note for the headpost - the 'Simplified' terrain types from the BMM should not be implemented as a different terrain type in the Map Editor - they are merely implemented differently in the rules when the 'Simplify Terrain (BMM)' option is switched on. If that option is off, they are the standard terrain from TW/TO.

**Comment by Ursur1minor** [comment]
Suggested addition to the list of terrain features from the _Battle of Tukayyid_ 

<img width="446" height="151" alt="Image" src="https://github.com/user-attachments/assets/9f5e143c-2b5b-47aa-a17c-d66c7fc123e5" />
Unnamed in the rules but I suggest calling it a "Ramp"
Ramp: If movement enters or exits a hex via the exits of a road with this feature the unit travelling along it does not need to pay for elevation changes as it is assumed the path is constructed in such a way to ease transition between different levels, only works if elevation differences are no more than 1

**Comment by HammerGS** [comment]
The **Water Flow** line on this list now has a dedicated issue with the rules and implementation notes written up: #8787.

Worth recording what turned out to be already done, because it is more than the list implies. Rapids and torrents themselves are fully implemented - the terrain, both levels, the movement cost (+1 and +2 MP), the piloting modifier (+2 and +3), board editor support, hex validation, creation by weather above Moderate Gale and Storm, and placement by the random map generator. That part was #342 and it was correctly closed.

What is missing is only the displacement half: nothing in the codebase has a direction of flow, and no unit is ever carried downstream. That is what #8787 covers.
